File: M/1870-minimum-speed-to-arrive-on-time/main.go
package main

import "fmt"

func minSpeedOnTime(dist []int, hour float64) int {
	if hour <= float64(len(dist)-1) {
		return -1
	}

	min, max := 1, 1<<16-1
	for min <= max {
		speed := min + (max-min)/2
		var hours float64 = 0.0
		for i := 0; i < len(dist)-1; i++ {
			h := dist[i] / speed
			if h*speed < dist[i] {
				h++
			}
			hours += float64(h)
		}
		hours += float64(dist[len(dist)-1]) / float64(speed)
		if hours <= hour {
			max = speed - 1
		} else {
			min = speed + 1
		}
	}
	return min
}

func main() {
	fmt.Println()
}

File: M/1870-minimum-speed-to-arrive-on-time/main.rs
struct Solution {}

impl Solution {
  pub fn min_speed_on_time(dist: Vec<i32>, hour: f64) -> i32 {
    let mut s: i32 = 1;
    let sum: i32 = dist.iter().sum::<i32>();
    let mut e: i32 = 1000000001;
    while s <= e {
      let m = s + (e - s) / 2;
      let mut sum: i32 = 0;
      let mut idx: usize = 0;
      while idx < dist.len() - 1 {
        sum += (dist[idx] + m - 1) / m;
        idx += 1;
      }
      let mut sum: f64 = sum as f64 + dist[dist.len() - 1] as f64 / m as f64;
      if sum <= hour {
        e = m - 1;
      } else {
        s = m + 1;
      }
    }
    match s > 1000000001 {
      true => -1,
      _ => s,
    }
  }
}
